coresimd/src/x86/i686/sse2.rs

```diff
@@ -25,56 +25,22 @@ pub unsafe fn _mm_cvtsi64x_sd(a: f64x2, b: i64) -> f64x2 {
     _mm_cvtsi64_sd(a, b)
 }
 
-/// Return a vector whose lowest element is `a` and all higher elements are
-/// `0`.
-#[inline(always)]
-#[target_feature = "+sse2"]
-// no particular instruction to test
-pub unsafe fn _mm_cvtsi64_si128(a: i64) -> i64x2 {
-    i64x2::new(a, 0)
-}
-
-/// Return a vector whose lowest element is `a` and all higher elements are
-/// `0`.
-#[inline(always)]
-#[target_feature = "+sse2"]
-// no particular instruction to test
-pub unsafe fn _mm_cvtsi64x_si128(a: i64) -> i64x2 {
-    _mm_cvtsi64_si128(a)
-}
-
-/// Return the lowest element of `a`.
-#[inline(always)]
-#[target_feature = "+sse2"]
-// no particular instruction to test
-pub unsafe fn _mm_cvtsi128_si64(a: i64x2) -> i64 {
-    a.extract(0)
-}
-
-/// Return the lowest element of `a`.
-#[inline(always)]
-#[target_feature = "+sse2"]
-// no particular instruction to test
-pub unsafe fn _mm_cvtsi128_si64x(a: i64x2) -> i64 {
-    _mm_cvtsi128_si64(a)
-}
-
 /// Returns the lower 64 bits of a 128-bit integer vector as a 64-bit
 /// integer.
 #[inline(always)]
 #[target_feature = "+sse2"]
-// no particular instruction to test
-pub unsafe fn _mm_movepi64_pi64(a: i64x2) -> i64 {
-    a.extract(0)
+// #[cfg_attr(test, assert_instr(movdq2q))] // FIXME: llvm codegens wrong instr?
+pub unsafe fn _mm_movepi64_pi64(a: i64x2) -> i8x8 {
```

Comment:
This is wrong, this function returns a single 64-bit integer. It should have been `__m64` instead of `i64` though (probably `__m64` wasn't implemented back then).

We should go over all intrinsics that return `__m64` and make sure that we don't make them worse by returning something else. In particular, chaining one intrinsic that returns `__m64` with one that takes `__m64` as an argument probably generates assembly that operates directly on that MMX register. However, `u8x8` is not an `x86_mmx` LLVM type, so by using a different type that does not translates to `x86_mmx` we might be in-curing a cost that the C compilers don't have.